src/main/java/javaslang/control/Apply.java

```diff
@@ -0,0 +1,15 @@
+/*     / \____  _    _  ____   ______  / \ ____  __    _ _____
+ *    /  /    \/ \  / \/    \ /  /\__\/  //    \/  \  / /  _  \   Javaslang
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/  \__/  /   Copyright 2014-now Daniel Dietrich
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/___\_/  \_//  \__/_____/    Licensed under the Apache License, Version 2.0
+ */
+package javaslang.control;
+
+import javaslang.Function1;
+import javaslang.algebra.Functor;
+
+public interface Apply<T> extends Functor<T> {
+
+    <U> Apply<U> ap(Apply<Function1<T,U>> f);
```

Comment:
In Javaslang we choose readable methods. Otherwise our code gets too cryptic. Let's call it `apply` instead of `ap`.

Instead of Function1 we use Function in the case of method parameter type (Function1 extends Function).

As you mentioned already, here we have the same problem as with Monad.flatMap because of the lack of higher-kinded types / type constructors. We circumvented this in Monad.flatMap by having a Function arg that returns Iterator.

Here I would give this signature a try:

```java
<U> Applicative<U> apply(Applicative<? extends Function<? super T, ? extends U>> f);
```

Question: How will implementations deal with an `f` that has a different Applicative type than the present?

Alternative: We could add a _recursive_ self-type `Applicative<SELF extends Applicative<SELF, ?>, T>` which is not a good idea.